authentication: add service account and webhook docs and update formatting

ericchiang opened the pull request:
It was brought up in #sig-auth today that the service account authenticator isn't mentioned in the authentication docs, and that this document is generally pretty outdated. Took the opportunity to fix up the following:

* Added top level section about what authenticators do.
* Added section about service account tokens.
* Added section describing webhook authenticator.
* Updated OpenID Connect section.
* Provided more recent link to Keystone authenticator work.
* Generally updated formatting to bring more in line with the authorizer document.
* Removed @yifan-gu from assignees as I'm taking over the OpenID Connector authenticator from him.

cc @kubernetes/sig-auth @lavalamp @erictune 

edit:

This possibly addresses #598 @nVitius

lavalamp commented:
@deads2k

pwittrock commented:
@lavalamp should @deads2k be the reviewer?

lavalamp commented:
YEs.

On Wed, Aug 3, 2016 at 4:09 PM, Phillip Wittrock <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> should @deads2k
> <https://github.com/deads2k> be the reviewer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes.github.io/pull/959#issuecomment-237402992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglo4q6EcCjTu_TAeNjPvVilb2xnrJks5qcR-qgaJpZM4JcJqw>
> .
>

pwittrock commented:
@ericchiang Please also add @deads2k as an assignee

deads2k commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+* `--oidc-username-claim` JWT claim to use as the user name. By default `sub`,
+which is expected to be a unique identifier of the end user. Admins can choose
+other claims, such as `email`, depending on their provider.
+* `--oidc-groups-claim` JWT claim to use as the user's group. If the claim is present
```
This is still listed as experimental in the API server and I'm not aware of pinned mapping.  Same with user.

ericchiang replied:
Thanks, I'll mark them as experimental.

liggitt commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+
+The plugin also accepts the following optional flags:
+
+* `--oidc-ca-file` Used by the apiserver to establish and verify the secure connection to the issuer.
```
if omitted, it uses system roots... works for OIDC servers with certs signed by public CAs

deads2k commented on docs/admin/authentication.md:
````diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+}
+```
+
+HTTP status codes can be used to supply additional erro context.
````
typo

deads2k commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+* `--oidc-groups-claim` JWT claim to use as the user's group. If the claim is present
+it must be an array of strings.
+
+## Webhook Token Authentication
```
Have you actually wired this up?  It was my intent with https://github.com/kubernetes/kubernetes/pull/29796, but I've only actually wired it in openshift for node authentication.

ericchiang replied:
No. I'll run an experiment today to see if these docs actually work.

ericchiang replied:
I was able to successfully get this working with a 1.3.4 cluster. Only typo is "tokenReview" should be "TokenReview"

erictune commented on docs/admin/authentication.md:
```diff
@@ -2,21 +2,40 @@
-**Client certificate authentication** is enabled by passing the `--client-ca-file=SOMEFILE`
+* Username: a string which identifies the end user. Common values might be `kube-admin` or `jane@example.com`.
+* UID: a string which identifies the end user and attempts to be more consistent and unique than username.
+* Groups: a list of strings which associate users with as set of commonly grouped users.
```
a set

erictune commented on docs/admin/authentication.md:
```diff
@@ -2,21 +2,40 @@
 option to apiserver. The referenced file must contain one or more certificates authorities
 to use to validate client certificates presented to the apiserver. If a client certificate
 is presented and verified, the common name of the subject is used as the user name for the
 request.
```
See   APPENDIX for how to generate a client cert.

erictune commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
 
-**Keystone authentication** is enabled by passing the `--experimental-keystone-url=<AuthURL>`
+## Service Account Tokens
+
```
Put this somewhere, maybe at the top:  

All Kubernetes clusters have two categories of users.  

1. Service Accounts are accounts that are managed by Kubernetes:
  - some are automatically created, but they can also be manually created
  - they are associated with a single Kubernetes cluster and namespace
  - they are automatically deleted when a namespace is deleted
  - credentials are automatically generated for service accounts
  - credentials are automatically mounted in volumes for Pods. 
  - They are intended to be used by programs running in Pods.
  - Service account names are scoped to namespaces.
2. Normal users are not managed by Kubernetes
  - The canonical storage for normal users is assumed to be outside Kubernetes.
  - The lifecycle for normal users is assumed to be independent of Kubernetes clusters and objects.
  - Normal user names are not scoped to a cluster or namespace.
  - Kubernetes does not have objects which represent normal user accounts.
  - Kubernetes authenticators identify the user who is making a request.

erictune commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+
+When a client attempts to authenticate with the apiserver using a bearer token,
+the authentication webhook queries the remote service with a review object
+containing the token.
```
The client has to provide an `Authorization: Bearer $TOKEN` HTTP header.  Kubernetes will not challenge request that lack such a header.

erictune commented on docs/admin/authentication.md:
```diff
@@ -60,19 +59,213 @@ more secure modes described above easier to use.
+containing the token.
+
+Note that webhook API objects are subject to the same [versioning compatibility rules](/docs/api/)
+as other Kubernetes API objects. Implementers should be aware of loser
```
looser

erictune commented:
Nice improvements, Eric, thanks!

deads2k commented:
@ericchiang friendly bump.

ericchiang commented:
Fixed typos and resolved some smaller issues pointed out by @erictune and @deads2k.

Added a section at the top titled "Users in Kubernetes" to emphasize how Kubernetes thinks of users and the difference between service accounts and normal users.

Expanded the service accounts section to note that they're automatically created and associated with pods by the ServiceAccount Admission Controller.

ericchiang commented:
Oh couple questions. How do we spell:

* service account vs Service Account
* API server, API Server, apiserver?

deads2k commented:
> Oh couple questions. How do we spell:

I do "service account" and "API server", but I suspect that an editor will sweep eventually.

Changes lgtm.

ericchiang commented:
> I do "service account" and "API server", but I suspect that an editor will sweep eventually.

Okay, I've standardized on those for now and will defer to a future editor.

deads2k commented:
lgtm, manually merging per: https://github.com/kubernetes/kubernetes.github.io/pull/946#issuecomment-238952955
